File: C++/src/_00208_ImplementTrie.cpp
#include <string>
#include <map>

/**
 * A trie (pronounced as "try") or prefix tree is a tree data structure used
 * to efficiently store and retrieve keys in a dataset of strings. There are
 * various applications of this data structure, such as autocomplete and
 * spellchecker.
 */
class Trie
{
private:
    /**
     * Represents a character.
     */
    struct TrieNode
    {
        /** Characters that come immidiately after this character. */
        TrieNode *children[26]{nullptr};
        /** `true` if this character is the end of a word. */
        bool isEndOfWord = false;

        ~TrieNode()
        {
            for (int i = 0; i < 26; ++i)
            {
                delete children[i];
            }
        }
    };

    TrieNode *root;

public:
    /**
     * Initialize the trie object.
     */
    Trie()
    {
        root = new TrieNode();
    }

    ~Trie()
    {
        delete root;
    }

    /**
     * Insert the string `word` into the trie.
     * @param word the string to insert
     */
    void insert(std::string word)
    {
        TrieNode *curr = root;
        for (char c : word)
        {
            if (!curr->children[c - 'a'])
                curr->children[c - 'a'] = new TrieNode();
            curr = curr->children[c - 'a'];
        }
        curr->isEndOfWord = true;
    }

    /**
     * Return `true` if the string `word` is in the trie (i.e., was inserted
     * before), and `false` otherwise.
     * @param word the string to search for
     * @return `true` if the word is found or `false` otherwise
     */
    bool search(std::string word)
    {
        TrieNode *curr = root;

        for (char c : word)
        {
            if (!curr->children[c - 'a'])
                return false;
            curr = curr->children[c - 'a'];
        }

        return curr->isEndOfWord;
    }

    /**
     * Return `true` if there is a previously inserted string `word` that has
     * the prefix `prefix`, and `false` otherwise.
     * @param word the prefix to search for
     * @return `true` if a word has the `prefix` or `false` otherwise
     */
    bool startsWith(std::string prefix)
    {
        TrieNode *curr = root;

        for (char c : prefix)
        {
            if (!curr->children[c - 'a'])
                return false;
            curr = curr->children[c - 'a'];
        }

        return true;
    }
};

File: Python/_00208_ImplementTrie.py
class Trie:
    """
    A trie (pronounced as "try") or **prefix tree** is a tree data structure
    used to efficiently store and retrieve keys in a dataset of strings. There
    are various applications of this data structure, such as autocomplete and
    spellchecker.
    """

    class _TrieNode:
        """
        Node representing a character.

        Attributes:
            children (dict): list of characters that can immediately follow 
                this character
            isEndOfWord (bool): `True` is this character is the end of a word or
                `False` otherwise

        """
        def __init__(self):
            self.children: dict = {}
            self.isEndOfWord: bool = False


    def __init__(self):
        """
        Initialize the trie object.
        """
        self.root = Trie._TrieNode()

    def insert(self, word: str) -> None:
        """
        Insert the string `word` into the trie.

        Args:
            word (str): the string to insert
        """
        curr = self.root

        for c in word:
            if c not in curr.children:
                new_node = Trie._TrieNode()
                curr.children[c] = new_node

            curr = curr.children[c]

        curr.isEndOfWord = True

    def search(self, word: str) -> bool:
        """
        Returns `True` if the string `word` is in the trie (i.e., was inserted
        before), and `False` otherwise.

        Args:
            word (str): the string to search for

        Returns:
            bool: `True` if the string `word` is in the trie or `False` otherwise.
        """

        curr = self.root

        for c in word:
            if c not in curr.children:
                return False

            curr = curr.children[c]

        return curr.isEndOfWord

    def startsWith(self, prefix: str) -> bool:
        """
        Returns `True` if there is a previously inserted string `word` that has
        the prefix `prefix`, and `false` otherwise.

        Args:
            prefix (str): the string prefix to search for

        Returns:
            bool: `True` if there is a string with the `prefix` or `False`
            otherwise
        """

        curr = self.root

        for c in prefix:
            if c not in curr.children:
                return False

            curr = curr.children[c]

        return True


import unittest


class TestTrie(unittest.TestCase):
    def test_example1(self):
        trie = Trie()
        trie.insert("apple")
        self.assertTrue(trie.search("apple"))
        self.assertFalse(trie.search("app"))
        self.assertTrue(trie.startsWith("app"))
        trie.insert("app")
        self.assertTrue(trie.search("app"))

if __name__ == "__main__":
    unittest.main()
